### PR TITLE
Implemented ignores

### PR DESCRIPTION
**Pull request by Meldiron**
- [x] appwrite.json ("ignore" attribute)
- [x] .appwriteignore
- [x] .gitignore
- [x] Manual QA:

---

- `node_modules` in `appwrite.json`:
CleanShot 2022-03-29 at 10 26 33

- `node_modules` in `.appwriteignore`:
CleanShot 2022-03-29 at 10 26 46

- `index.js` in `.gitignore`:
CleanShot 2022-03-29 at 10 26 38

- `null` in `appwrite.json`, ano no ingnore files:
CleanShot 2022-03-29 at 10 26 51

### CODE REVIEW

File: templates/cli/lib/commands/command.js.twig [hunk 1]
```diff
@@ -37,13 +40,33 @@ const {{ service.name | caseLower }}{{ method.name | caseUcfirst }} = async ({ {
...
+    } else if(fs.existsSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.appwriteignore'))) {
+        ignorer.add(fs.readFileSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.appwriteignore')).toString());
+        log('Ignoring files using configuration from .appwriteignore');
+    } else if(fs.existsSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.gitignore'))) {
+        ignorer.add(fs.readFileSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.gitignore')).toString());
+        log('Ignoring files using configuration from .gitignore');
+    } else {
+        log('Not ignoring any files.');
+        // No ignoring in place
+    }
```
Comment (christyjacob4) [inline templates/cli/lib/commands/command.js.twig:59]:
```suggestion
    }
```

**Review by christyjacob4 — APPROVED** [review]

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Some comments

File: templates/cli/lib/commands/command.js.twig [hunk 2]
```diff
@@ -37,13 +40,33 @@ const {{ service.name | caseLower }}{{ method.name | caseUcfirst }} = async ({ {
...
         throw new Error('The path is not a directory.');
     
+    const ignorer = ignore();
+
+    const xfunction = localConfig.getFunction(functionId);
+
+    if(xfunction.ignore) {
+        ignorer.add(xfunction.ignore);
+        log('Ignoring files using configuration from appwrite.json');
+    } else if(fs.existsSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.appwriteignore'))) {
```
[the remark this answers is not in the document]
Comment (Meldiron) [reply templates/cli/lib/commands/command.js.twig]: Removed .appwriteignore logic

File: templates/cli/lib/commands/command.js.twig (hunk 1 above)
Comment (Meldiron) [reply templates/cli/lib/commands/command.js.twig]: Removed "else" log

File: templates/cli/lib/commands/command.js.twig [hunk 3]
```diff
@@ -37,13 +40,27 @@ const {{ service.name | caseLower }}{{ method.name | caseUcfirst }} = async ({ {
     if (!fs.lstatSync(folderPath).isDirectory())
         throw new Error('The path is not a directory.');
     
+    const ignorer = ignore();
+
+    const xfunction = localConfig.getFunction(functionId);
```
Comment (christyjacob4) [inline templates/cli/lib/commands/command.js.twig:45]:
```suggestion
    const func = localConfig.getFunction(functionId);
```

In deploy.js we're using `func` so let's try to be consistent. 

We need to either change both occurances to `func` or `xfunction`

Can you check whats a more widely accepted / accurate convention ?

File: templates/cli/lib/commands/command.js.twig [hunk 4]
```diff
@@ -37,13 +40,27 @@ const {{ service.name | caseLower }}{{ method.name | caseUcfirst }} = async ({ {
...
+
+    if(xfunction.ignore) {
+        ignorer.add(xfunction.ignore);
+        log('Ignoring files using configuration from appwrite.json');
+    } else if(fs.existsSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.gitignore'))) {
+        ignorer.add(fs.readFileSync(pathLib.join({{ parameter.name | caseCamel | escapeKeyword }}, '.gitignore')).toString());
+        log('Ignoring files using configuration from .gitignore');
```
Comment (christyjacob4) [inline templates/cli/lib/commands/command.js.twig:52]:
```suggestion
        log('Ignoring files in .gitignore');
```

File: templates/cli/lib/commands/command.js.twig [hunk 5]
```diff
@@ -37,13 +40,27 @@ const {{ service.name | caseLower }}{{ method.name | caseUcfirst }} = async ({ {
     if (!fs.lstatSync(folderPath).isDirectory())
         throw new Error('The path is not a directory.');
     
+    const ignorer = ignore();
+
+    const xfunction = localConfig.getFunction(functionId);
+
+    if(xfunction.ignore) {
+        ignorer.add(xfunction.ignore);
+        log('Ignoring files using configuration from appwrite.json');
```
Comment (christyjacob4) [inline templates/cli/lib/commands/command.js.twig:49]:
```suggestion
        log('Ignoring files using configuration from appwrite.json');
```

**Review by christyjacob4 — APPROVED** [review]
Let's make sure we update the docs as well.

File: templates/cli/lib/commands/command.js.twig (hunk 4 above)
Comment (Meldiron) [reply templates/cli/lib/commands/command.js.twig]: Also updated other message to `Ignoring files in appwrite.json`

**Review by PineappleIOnic — APPROVED** [review]

File: templates/cli/lib/questions.js.twig [hunk 6]
```diff
@@ -2,6 +2,31 @@ const { localConfig } = require('./config');
...
+    case 'php':
+      return ['vendor'];
+    case 'python':
+      return ['__pypackages__'];
+    case 'ruby':
+      return ['vendor'];
+    case 'rust':
+      return ['target', 'debug', '*.rs.bk', '*.pdb'];
+    case 'swift':
+      return ['Packages'];
```
Comment (abnegate) [inline templates/cli/lib/questions.js.twig:24]:
```suggestion
      return ['.build', '.swiftpm'];
```

File: templates/cli/lib/questions.js.twig [hunk 7]
```diff
@@ -2,6 +2,31 @@ const { localConfig } = require('./config');
...
 
+const getIgnores = (runtime) => {
+  const languge = runtime.split('-')[0];
+
+  switch (languge) {
+    case 'dart':
+      return ['.packages', '.dart_tool'];
+    case 'deno':
+      return [];
+    case 'node':
```
Comment (abnegate) [inline templates/cli/lib/questions.js.twig:13]:
```suggestion
    case 'dotnet':
        return ['bin', 'obj', '.nuget'];
    case 'java':
    case 'kotlin':
        return ['build'];
    case 'node':
```

File: templates/cli/lib/questions.js.twig [hunk 8]
```diff
@@ -2,6 +2,31 @@ const { localConfig } = require('./config');
 const { projectsList } = require('./commands/projects');
 const { functionsListRuntimes } = require('./commands/functions');
 
+const getIgnores = (runtime) => {
+  const languge = runtime.split('-')[0];
+
+  switch (languge) {
+    case 'dart':
```
Comment (abnegate) [inline templates/cli/lib/questions.js.twig:9]:
```suggestion
    case 'cpp':
        return ['build', 'CMakeFiles', 'CMakeCaches.txt'];
    case 'dart':
```

**Review by abnegate — CHANGES_REQUESTED** [review]

**Comment by christyjacob4** [comment]
@Meldiron  please fix the conflicts

**Pull request merged**